glide-core/redis-rs/redis/src/cluster_async/mod.rs

```diff
@@ -1371,62 +1372,154 @@ where
+                    Ok(node) => {
+                        info!(
+                            "Succeeded to refresh connection for node {}.",
+                            address_clone_for_task
+                        );
+                        inner_clone
+                            .conn_lock
+                            .write()
+                            .expect(MUTEX_READ_ERR)
+                            .replace_or_add_connection_for_address(&address_clone_for_task, node);
+                    }
+                    Err(err) => {
+                        warn!(
+                            "Failed to refresh connection for node {}. Error: `{:?}`",
+                            address_clone_for_task, err
+                        );
+                    }
+                }
+
+                // Note!! - TODO
+                // Need to notify here the awaiting requests inorder to awake the context of the poll_flush as
+                // it awaits on this notifier inside the get_connection in the poll_next inside poll_complete.
+                // Otherwise poll_flush won't be polled until the next start_send or other requests I/O.
+                inner_clone
+                    .conn_lock
+                    .write()
+                    .expect(MUTEX_READ_ERR)
+                    .refresh_conn_state
+                    .refresh_address_in_progress
+                    .remove(&address_clone_for_task);
+
+                debug!(
+                    "Refreshing connection task to {:?} is done",
+                    address_clone_for_task
+                );
             });
-        }
 
-        // Poll connection tasks as soon as each one finishes
-        while let Some(result) = tasks.next().await {
-            match result {
-                (address, Ok(node)) => {
-                    let connections_container = inner.conn_lock.read().expect(MUTEX_READ_ERR);
-                    connections_container.replace_or_add_connection_for_address(address, node);
-                }
-                (address, Err(err)) => {
-                    warn!(
-                        "Failed to refresh connection for node {}. Error: `{:?}`",
-                        address, err
-                    );
-                }
+            // Create the RefreshTaskState before spawning the task
+            let refresh_task_state = RefreshTaskState::new(handle);
+
+            // Extract the notifier from the new RefreshTaskState
+            if let RefreshTaskStatus::Reconnecting(ref notifier) = refresh_task_state.status {
+                notifiers.push(notifier.get_notifier());
             }
+
+            // Keep the task handle into the RefreshState of this address
+            inner
+                .conn_lock
+                .write()
+                .expect(MUTEX_READ_ERR)
+                .refresh_conn_state
+                .refresh_address_in_progress
+                .insert(address.clone(), refresh_task_state);
```

Comment:
It's confusing, I would expect you to explicitly create Reocnnection state here:

```suggestion
            let notifier = RefreshTaskNotifier::new();
            notifiers.push(notifier.get_notifier());

            // Keep the task handle into the RefreshState of this address
            inner
                .conn_lock
                .write()
                .expect(MUTEX_READ_ERR)
                .refresh_conn_state
                .refresh_address_in_progress
                .insert(
                    address.clone(),
                    RefreshTaskState {
                        handle,
                        status: RefreshTaskStatus::Reconnecting(notifier),
                    },
                );
```